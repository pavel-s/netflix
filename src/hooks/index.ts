export { useAuth } from './use-auth-user';
export { useAuthUser } from './use-auth-user';
export { useContent } from './use-content';
