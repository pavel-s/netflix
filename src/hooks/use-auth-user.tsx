import { useEffect, useState } from 'react';
import firebase, { auth, TFirebaseUser } from '../lib/firebase.prod';

const USER_STORAGE_KEY = 'authUser';

export const useAuthUser = () => {
  const storedUser = localStorage.getItem(USER_STORAGE_KEY);

  const [user, setUser] = useState<TFirebaseUser | null>(
    storedUser ? JSON.parse(storedUser) : null
  );

  useEffect(() => {
    const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
      if (user) {
        localStorage.setItem(USER_STORAGE_KEY, JSON.stringify(user));
        setUser(user);
      } else {
        localStorage.removeItem(USER_STORAGE_KEY);
        setUser(null);
      }
    });
    return unsubscribe;
  }, []);

  return user;
};

const signUp = ({
  name,
  email,
  password,
}: {
  name: string;
  email: string;
  password: string;
}): Promise<void> => {
  return auth.createUserWithEmailAndPassword(email, password).then((result) =>
    result.user?.updateProfile({
      displayName: name,
      photoURL: `images/users/${Math.floor(Math.random() * 5) + 1}.png`,
    })
  );
};

const signOut = () => auth.signOut();

const signIn = ({ email, password }: { email: string; password: string }) => {
  return auth.signInWithEmailAndPassword(email, password);
};

const authMethods = { signUp, signOut, signIn };
export const useAuth = () => {
  return authMethods;
};
