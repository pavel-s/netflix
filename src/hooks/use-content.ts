import { useEffect, useState } from 'react';
import firebase from '../lib/firebase.prod';
import { TMediaItem } from '../types';

export const useContent = (type: 'films' | 'series') => {
  const [content, setContent] = useState<TMediaItem[]>([]);
  useEffect(() => {
    firebase
      .firestore()
      .collection(type)
      .get()
      .then((snapshot) => {
        const result = snapshot.docs.map(
          (docObj) =>
            ({
              ...docObj.data(),
              docId: docObj.id,
            } as TMediaItem)
        );
        setContent(result);
      })
      .catch((error) => console.log(error.message));
  }, [type]);
  return content;
};
