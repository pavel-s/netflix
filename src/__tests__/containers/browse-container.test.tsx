import { act, fireEvent, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import BrowseContainer from '../../containers/browse';
import { useAuth, useContent } from '../../hooks';
import content from './mockContent.json';

const profile = {
  displayName: 'Bob',
  photoURL: '/images/users/1.png',
};

jest.mock('../../hooks', () => {
  return {
    useContent: jest.fn(),
    useAuth: jest.fn(),
  };
});

describe('<BrowseContainer />', () => {
  it('renders BrowseContainer with populated data', async () => {
    // @ts-ignore
    useContent.mockImplementation((category) => content[category]);
    // @ts-ignore
    useAuth.mockReturnValue({
      signOut: () => Promise.resolve('sign out success'),
    });

    const { container, getByText, getByAltText, getByTestId, queryByAltText } =
      render(<BrowseContainer profile={profile} />, {
        wrapper: MemoryRouter,
      });

    // categories
    expect(getByText('Films')).toBeTruthy();
    expect(getByText('Series')).toBeTruthy();

    // item from category 'Films'
    expect(getByAltText('Kings Speech')).toBeTruthy();

    // toggle category to 'Series'
    await act(async () => {
      fireEvent.click(getByText('Series'));
    });

    expect(useContent).toHaveBeenCalledWith('series');

    // item from category 'Series'
    expect(getByAltText('The Office')).toBeTruthy();

    // search with term 'The'
    await act(async () => {
      fireEvent.change(getByTestId('search-input'), {
        target: { value: 'The' },
      });
    });

    // 'The Office' found
    expect(getByAltText('The Office')).toBeTruthy();

    // search with term 'juno'
    await act(async () => {
      fireEvent.change(getByTestId('search-input'), {
        target: { value: 'juno' },
      });
    });

    // 'The Office' not found, but 'Juno' found
    expect(queryByAltText('The Office')).toBeFalsy();
    expect(queryByAltText('Juno')).toBeTruthy();

    // card feature hidden
    expect(queryByAltText('close')).toBeFalsy();

    // click on 'Juno' card
    await act(async () => {
      fireEvent.click(getByAltText('Juno'));
    });

    // card feature show (close button)
    expect(queryByAltText('close')).toBeTruthy();

    // close card feature
    await act(async () => {
      fireEvent.click(getByAltText('close'));
    });

    // card feature hidden
    expect(queryByAltText('close')).toBeFalsy();

    // clear search and toggle back to films
    await act(async () => {
      fireEvent.change(getByTestId('search-input'), { target: { value: '' } });
      fireEvent.click(getByText('Films'));
    });

    // item from category 'Films'
    expect(getByAltText('Kings Speech')).toBeTruthy();

    expect(container).toMatchSnapshot();
  });

  it('renders BrowserContainer with profile without photo', () => {
    // @ts-ignore
    useContent.mockImplementation((category) => content[category]);
    // @ts-ignore
    useAuth.mockReturnValue({
      signOut: () => Promise.resolve('sign out success'),
    });

    const { queryByText } = render(
      <BrowseContainer profile={{ ...profile, photoURL: null }} />,
      { wrapper: MemoryRouter }
    );

    expect(queryByText('Bob')).toBeTruthy();
  });
});
