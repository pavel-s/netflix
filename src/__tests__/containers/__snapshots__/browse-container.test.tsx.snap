// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrowseContainer /> renders BrowseContainer with populated data 1`] = `
<div>
  <div
    class="header__Background-sc-1qf3jwc-0 ioQOEo"
  >
    <div
      class="header__Container-sc-1qf3jwc-1 iJrjfW"
    >
      <div
        class="header__Group-sc-1qf3jwc-9 fRouHy"
      >
        <a
          class="header__LogoLink-sc-1qf3jwc-3 eTWTRb"
          href="/"
        >
          <img
            class="header__Logo-sc-1qf3jwc-2 edGsUc"
            src="images/misc/logo.svg"
          />
        </a>
        <button
          class="header__TextButton-sc-1qf3jwc-8 fCFpgM"
        >
          Films
        </button>
        <button
          class="header__TextButton-sc-1qf3jwc-8 CpwQL"
        >
          Series
        </button>
      </div>
      <div
        class="header__Group-sc-1qf3jwc-9 fRouHy"
      >
        <div
          class="header__Search-sc-1qf3jwc-15 gKHaVv"
        >
          <button
            data-testid="search-button"
          >
            <img
              class="header__SearchIcon-sc-1qf3jwc-13 gFOGps"
              src="images/icons/search.png"
            />
          </button>
          <input
            class="header__SearchInput-sc-1qf3jwc-14 dCGTSp"
            data-testid="search-input"
            type="text"
            value=""
          />
        </div>
        <div
          class="header__Profile-sc-1qf3jwc-12 dJSJdG"
        >
          <div
            class="header__ProfileImage-sc-1qf3jwc-10 dxgfkT"
          />
          <div
            class="header__Dropdown-sc-1qf3jwc-11 eGVzkl"
          >
            <div
              class="header__Group-sc-1qf3jwc-9 fRouHy"
            >
              <div
                class="header__ProfileImage-sc-1qf3jwc-10 dxgfkT"
              />
              <button
                class="header__TextButton-sc-1qf3jwc-8 CpwQL"
              >
                Bob
              </button>
            </div>
            <button
              class="header__TextButton-sc-1qf3jwc-8 CpwQL"
            >
              Sign out
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="header__Feature-sc-1qf3jwc-5 YbyHX"
    >
      <h2
        class="header__FeatureCallOut-sc-1qf3jwc-6 eUrcDy"
      >
        Watch Joker now.
      </h2>
      <p
        class="header__Text-sc-1qf3jwc-7 eVXLUY"
      >
        Forever alone in a crowd, failed comedian Arthur Fleck seeks connection as he walks the streets of Gotham City. Arthur wears two masks -- the one he paints for his day job as a clown, and the guise he projects in a futile attempt to feel like he's part of the world around him.
      </p>
      <button
        class="header__PlayButton-sc-1qf3jwc-16 TubXA"
      >
        Play
      </button>
    </div>
  </div>
  <div
    class="card__Container-vait2u-0 gDjoKe"
  >
    <div
      class="card__CardContainer-vait2u-1 csYGmB"
    >
      <h2
        class="card__Title-vait2u-2 fmrCPZ"
      >
        Drama
      </h2>
      <div
        class="card__Items-vait2u-3 grIBiZ"
      >
        <button
          class="card__Item-vait2u-6 iMYRJV"
        >
          <img
            alt="Kings Speech"
            class="card__ItemImage-vait2u-4 ehhpt"
            src="/images/films/drama/kings-speech/small.jpg"
          />
          <div
            class="card__Meta-vait2u-5 hzdAah"
          >
            <h3
              class="card__Subtitle-vait2u-7 ejogZR"
            >
              Kings Speech
            </h3>
            <p
              class="card__Text-vait2u-8 hUQirf"
            >
              King George VI tries to overcome his stammering problem with the help of speech therapist Lionel Logue and makes himself worthy enough to lead his country through World War II.
            </p>
          </div>
        </button>
      </div>
    </div>
  </div>
  <div
    class="footer__Container-sc-58hhi-0 gLeidq"
  >
    <div
      class="footer__Inner-sc-58hhi-1 hxpWbN"
    >
      <p
        class="footer__Title-sc-58hhi-2 iYvwUS"
      >
        Questions? Call 
        <a
          href="tel:8-800-100-9668"
        >
          8-800-100-9668
        </a>
      </p>
      <ul
        class="footer__Links-sc-58hhi-3 dVUMzK"
      >
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            FAQ
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Help Center
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Account
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Media Center
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Investor Relations
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Jobs
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Ways to Watch
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Terms of Use
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Privacy
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Cookie Preferences
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Corporate Information
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Contact Us
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Speed Test
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Legal Notices
          </a>
        </li>
        <li
          class="footer__Link-sc-58hhi-4 gBCxki"
        >
          <a
            href="#"
          >
            Netflix Originals
          </a>
        </li>
      </ul>
      <p
        class="footer__Text-sc-58hhi-5 eNZVhR"
      >
        Netflix Russia
      </p>
    </div>
  </div>
</div>
`;
